hr_planning_resources/models/hr_leave.py

```diff
@@ -0,0 +1,181 @@
+            leaves[leave.employee_id.id].append(leave)
+        return leaves
+
+    def _get_leave_warning(self, leaves, employee, date_from, date_to):
+        loc_cache = {}
+
+        def localize(date):
+            if date not in loc_cache:
+                loc_cache[date] = (
+                    utc.localize(date)
+                    .astimezone(timezone(self.env.user.tz or "UTC"))
+                    .replace(tzinfo=None)
+                )
+            return loc_cache.get(date)
+
+        warning = ""
+        periods = self._group_leaves(leaves, employee, date_from, date_to)
+        periods_by_states = [
+            list(b) for a, b in groupby(periods, key=lambda x: x["is_validated"])
+        ]
+
+        for periods in periods_by_states:
+            period_leaves = ""
+            for period in periods:
+                dfrom = period["from"]
+                dto = period["to"]
+                prefix = ""
+                if period != periods[0]:
+                    if period == periods[-1]:
+                        prefix = _(" and")
+                    else:
+                        prefix = ","
+
+                if period.get("show_hours", False):
+                    period_leaves += _(
+                        "%(prefix)s from the %(dfrom_date)s at %(dfrom)s to the %(dto_date)s at %(dto)s",
+                        prefix=prefix,
+                        dfrom_date=format_date(self.env, localize(dfrom)),
+                        dfrom=format_time(self.env, localize(dfrom)),
+                        dto_date=format_date(self.env, localize(dto)),
+                        dto=format_time(self.env, localize(dto)),
+                    )
+                else:
+                    period_leaves += _(
+                        "%(prefix)s from the %(dfrom)s to the %(dto)s",
+                        prefix=prefix,
+                        dfrom=format_date(self.env, localize(dfrom)),
+                        dto=format_date(self.env, localize(dto)),
+                    )
+
+            time_off_type = (
+                _("is on time off")
+                if periods[0].get("is_validated")
+                else _("has requested time off")
+            )
+            warning += _(
+                "%(employee)s %(time_off_type)s%(period_leaves)s. \n",
+                employee=employee.name,
+                period_leaves=period_leaves,
+                time_off_type=time_off_type,
+            )
+        return warning
+
+    def _group_leaves(self, leaves, employee_id, date_from, date_to):
+        """
+        Returns all the leaves happening between `planned_date_begin` and `planned_date_end`
+        """
+        work_times = {
+            wk[0]: wk[1]
+            for wk in employee_id.list_work_time_per_day(date_from, date_to)
+        }
+
+        def has_working_hours(start_dt, end_dt):
+            """
+            Returns `True` if there are any working days between `start_dt` and `end_dt`.
```

Comment:
Should be imperative style https://stackoverflow.com/a/54868794
Same in previous docstring

Reply:
Thanks for the tips!!!